privacyidea/lib/tokens/certificatetoken.py

```diff
@@ -200,6 +409,27 @@ def update(self, param):
             self.add_tokeninfo("CA", ca)
             cacon = get_caconnector_object(ca)
         if request:
+            if not spkac:
+                # We only do the whole attestation checking in case we have no SPKAC
+                request_csr = load_pem_x509_csr(to_byte_string(request), default_backend())
+                if not request_csr.is_signature_valid:
+                    raise privacyIDEAError("request has invalid signature.")
+                # If a request is sent, we can have an attestation certificate
+                attestation = getParam(param, "attestation", optional)
+                verify_attestation = getParam(param, "verify_attestation", optional)
+                if attestation:
+                    request_numbers = request_csr.public_key().public_numbers()
+                    attestation_cert = load_pem_x509_certificate(to_byte_string(attestation), default_backend())
+                    attestation_numbers = attestation_cert.public_key().public_numbers()
+                    if request_numbers != attestation_numbers:
+                        log.warning("certificate request does not match attestation certificate.")
+                        raise privacyIDEAError("certificate request does not match attestation certificate.")
+                    if verify_attestation:
```

Comment:
Maybe we should always verify the certificate (and log a message and continue if it fails) and only fail hard if it is required by policy.

Reply:
Nice idea, will do this.

Reply:
Done in 8a02f887